querysql/querysql.go

```diff
@@ -133,39 +122,78 @@ func MustNextResult[T any](rs *ResultSets, typ func() Result[T]) T {
 	return result
 }
 
+func (rs *ResultSets) processAllLogSelects() error {
+	for !rs.Done() {
+		cols, err := rs.Rows.Columns()
+		if err != nil {
+			return err
+		}
+
+		if rs.hasLogColumn(cols) {
+			if err = rs.processLogSelect(); err != nil {
+				return err
+			}
+
+			if err = rs.nextResultSet(); err != nil {
+				return err
+			}
+		} else {
+			// non-logging select; return
+			return nil
+		}
+	}
+	return nil
+}
+
+func (rs *ResultSets) Done() bool {
+	return rs.Rows == nil
+}
+
+func (rs *ResultSets) nextResultSet() error {
+	if rs.Rows.NextResultSet() {
+		return nil
+	} else {
+		// we have exhausted the results; automatically close Rows; this also ensures Done() returns true
+		return rs.Close()
+	}
+}
+
 // Next reads the next result set from `rs`, passing each row to `scanner`;
 // taking care of checking errors and advancing result sets. On errors, `rs`
-// will be closed. If AutoClose is used, `rs` will also be closed on successful return.
+// will be closed. If EnsureDoneAfterNext is used, `rs` will also be closed on successful return.
 func Next(rs *ResultSets, scanner Target) error {
 	if rs.Err != nil {
 		return rs.Err
 	}
 
+	// When we enter this function one of the following will be the case:
+	// 1) We are at the beginning of using rs.Rows
+	// 2) rs.Rows.NextResultSet() was called, returned true, and there was another result set ready for us
+	// 3) rs.Rows.NextResultSet() was called, returned false, and we have done a close and rs.Rows is nil
+
 	success := false
 	defer func() {
 		if !success {
-			_ = rs.Rows.Close()
+			_ = rs.Close()
 		}
 	}()
 
-	for {
-
-		if !rs.BeginResultSet() {
-			return ErrNoMoreSets
-		}
-
-		cols, err := rs.Rows.Columns()
-		if err != nil {
+	if !rs.started {
+		// Case 1
```

Comment:
I don't understand that saying // Case 1 here is useful.

Reply:
```suggestion
```